apps/web/src/hooks/useWriteBaseEnsTextRecords.ts

```diff
@@ -1,85 +1,221 @@
+  const { logError } = useErrors();
+
+  // Fetch existing TextRecords
+  const { existingTextRecords, existingTextRecordsIsLoading, refetchExistingTextRecords } =
+    useReadBaseEnsTextRecords({
+      address,
+      username,
+    });
+
+  // Save a copy of updated TextRecords
+  const [updatedTextRecords, setUpdatedTextRecords] =
+    useState<UsernameTextRecords>(existingTextRecords);
+  useEffect(() => {
+    setUpdatedTextRecords(existingTextRecords);
+  }, [existingTextRecords]);
+
+  const updateTextRecords = useCallback(
+    (key: UsernameTextRecordKeys, value: string) => {
+      setUpdatedTextRecords((previousTextRecords) => {
+        return {
+          ...previousTextRecords,
+          [key]: value,
+        };
+      });
+    },
+    [setUpdatedTextRecords],
+  );
+
+  // Track keys that have changed
+  const keysToUpdate = useMemo(() => {
+    const keys: UsernameTextRecordKeys[] = Object.keys(
+      updatedTextRecords,
+    ) as UsernameTextRecordKeys[];
+    return keys.filter((key) => {
+      const existingValue = existingTextRecords[key].trim();
+      const updatedValueValue = updatedTextRecords[key].trim();
+
+      return existingValue != updatedValueValue;
+    });
+  }, [existingTextRecords, updatedTextRecords]);
+
+  // Track hasChanged as a boolean
+  const hasChanged = useMemo(() => {
+    return keysToUpdate.length !== 0;
+  }, [keysToUpdate.length]);
 
   const { basenameChain } = useBasenameChain();
 
+  // Write TextRecords
   const {
     data: writeTextRecordsTransactionHash,
     writeContractAsync,
     isPending: writeTextRecordsIsPending,
     isError: writeTextRecordsIsError,
     error: writeTextRecordsError,
+    isSuccess: writeTextRecordsIsSuccess,
+    reset: writeTextRecordsReset,
   } = useWriteContract();
 
+  // Wait for TextRecords transaction to be processed
+  const {
+    data: transactionData,
+    isFetching: transactionIsFetching,
+    isSuccess: transactionIsSuccess,
+    isError: transactionIsError,
+    error: transactionError,
+  } = useWaitForTransactionReceipt({
+    hash: writeTextRecordsTransactionHash,
+    chainId: basenameChain.id,
+    query: {
+      enabled: !!writeTextRecordsTransactionHash,
+    },
+  });
+
+  const { refetch: refetchBaseEnsAvatar } = useBaseEnsAvatar({
+    name: username,
+  });
+
   const { switchChainAsync } = useSwitchChain();
 
-  const writeTextRecords = useCallback(
-    async (textRecords: UsernameTextRecords) => {
-      if (!username) return Promise.reject(new Error('Cannot write text records without a name'));
-      const keys: UsernameTextRecordKeys[] = Object.keys(textRecords) as UsernameTextRecordKeys[];
-      const keysToUpdate = keys.filter((key) => {
-        const existingValue = existingTextRecords[key].trim();
-        const updatedValueValue = textRecords[key].trim();
+  const writeTextRecords = useCallback(async () => {
+    if (!username) return Promise.reject(new Error('Cannot write text records without a name'));
```

Comment:
I think you could use throw here instead

Reply:
I thought Promise.reject was [preferred for asynchronous function](https://stackoverflow.com/questions/33445415/javascript-promises-reject-vs-throw)?